allow tab indexes to be passed to the `tab-detach` command

gilbertw1 opened the pull request:
This pull request implements the feature requested in #2913

It allows individual tab indexes and ranges to be passed to the `tab-detach` command and opens the set of tabs in a new single window. If called with no tab indexes, it only detaches the current tab. Additionally, it verifies that at least one tab will remain open in the current window, and raises an error message otherwise.

This code borrows heavily from #2384/#2895 most notably the `parse_number_sets` function, as recommended in the issue.

The code is relatively straight forward. It parses out a set of tab indexes that are verified to be within the bounds of the existing tab indexes, or just the current tab if no indexes are provided. It makes a first pass through all the indexes to retrieve the tab widget and url for each tab to be detached. This first pass is made prior to removing any tabs so we won't have to worry about the tab indexes changing while we're walking through them. Finally we iterate through each tab/url tuple and open the url in a new window and close the tab widget in the current window.

A few things to note, that I could probably use guidance on:

* The most controversial change in this PR is updating the `_open` function to return a `TabbedBrowser`. I did this because I could find no other way to reference the newly created window to attach the rest of the detached tabs to. The only other way I could think of doing this was to actually reproduce the logic to create the window and new tab inline in the `tab-detach` function. However, this is probably not desirable as it would mean the logic (along with private tab/window handling) would be duplicated and hurt overall maintainability. I'm interested on what your thoughts are on this change.

* The previous implementation of `tab-detach` functioned the same way, but with this implementation it's much more noticeable that the tabs are loaded fresh after they are detached. Detaching five or so tabs results in a very noticeable refresh of all the tabs simultaneously. I was wondering if you think it'd be possible / advisable to implement a 'soft' tab remove function that would allow for tabs to be migrated from one `TabbedBrowser` to another. Looking at the underlying api's it appears to be possible to do, but I thought I'd run it by you first to see if you think it'd be a wise decision to migrate tabs. If this is possible, it'd be nice to implement a followup function that would allow tabs to be moved between qutebrowser windows.

I initially picked this issue up because I wanted to contribute to the project and it was marked as easy, but after implementing it, I'm very suprised at how useful it's turned out to be. I'll also mention that I've written an extremely small amount of Python, so don't be afraid to tear me a new one if I'm doing something bad here :)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/2937)
<!-- Reviewable:end -->

The-Compiler commented:
Thanks for the contribution and for the detailed explanation! 👍 

However, as mentioned in the [contribution guidelines](https://github.com/qutebrowser/qutebrowser/blob/master/CONTRIBUTING.asciidoc), I'm currently busy with exams, so reviewing this might take a bit longer. Also, the main focus after my exams is getting the new config in, so there might be some conflicts with the PR then.

In the meantime, could you maybe have a look at the [test/linter failures](https://travis-ci.org/qutebrowser/qutebrowser/builds/269823800) and also copy the tests for `NumberSet` from #2895? 

>  I was wondering if you think it'd be possible / advisable to implement a 'soft' tab remove function that would allow for tabs to be migrated from one TabbedBrowser to another.

Absolutely! The reason it wasn't done like that in the first place was because a cleaned up tab API like now didn't really exist yet, and it was easier like this. See #2636. That'd probably be a separate PR, though.

I'm also wondering whether any other `:tab-*` commands would benefit from taking a set of tabs like this, for consistency.

gilbertw1 commented:
Ah tests, right :)

I've pulled over and enhanced the test cases for the `parse_number_set` function along with the function itself. Additionally, I've added a bdd test for detaching multiple tabs. I also fixed up the test/linter failures. I noticed that the CI checks still failed, but it looks like that is due to other reasons?

I shied away from using the implementation of `NumberSet`, primarily because it didn't create an actual set of numbers and instead just stored a list of ranges, which could only be tested for existence. Additionally, it mostly ignores the `nummax` parameter and doesn't actually verify that the number being tested is within the specified bounds. I much prefer the concrete set creation for a number of reasons, and I think the built in set type does a great job providing all functions expected of a set.

Looking at the `:tab-*` functions, I think that: `tab-clone`, `tab-close`, `tab-pin`, and `tab-only` are good candidates for including this functionality. I'll update them to allow for sets to be provided as well and add those commits to this PR.

I'll also look into #2636 and see if I can assist in this effort.

Good luck on your exams!

The-Compiler commented:
The CI failures still seem to be related to your additions :wink: The [test environment](https://travis-ci.org/qutebrowser/qutebrowser/jobs/269969613#L1018) is failing because 100% test coverage is enforced for `utils.py`, and [pylint](https://travis-ci.org/qutebrowser/qutebrowser/jobs/269969624#L758)/[flake8](https://travis-ci.org/qutebrowser/qutebrowser/jobs/269969625#L574) are failing for various small style issues.

cc @Al-Caveman @Kingdread for the numberset stuff

gilbertw1 commented:
100% test coverage, now that's legit! Now that I understand the output better, I've made sure that all the checks pass.

Running the test suite got waaay less painful once I actually read that message about installing 'xvfb'.

Kingdread commented on qutebrowser/utils/utils.py:
```diff
@@ -856,3 +856,47 @@ def expand_windows_drive(path):
+        minnum: Minimum allowable number.
+        maxnum: Maximum allowable number.
+    """
+    txtset = txtset.lower() \
```
Not sure what @The-Compiler's preference is here, but [PEP-8](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) advises to do multi-line expressions via implicit continuations, i.e. wrapping the whole thing in parenthesis:

```python
txtset = (txtset.lower()
    .replace(...)
    .replace(...))
```

gilbertw1 replied:
I'd be happy to make this change. I've written very little Python, so I'm picking up the conventions as I go and have been relying on pylint to steer me the right direction. I imagine if the official style guide recommends parenthesis, then that's probably the way to go.

The-Compiler replied:
I agree!

Kingdread commented:
The numberset thing looks good to me. Though if more commands get the ability to use indexsets, it might be worth putting the documentation

> Specify the set of download indexes to delete. E.g. '2,5,6-last,3', where 'last' is special keyword that denotes that latest download index. If you specify numbers smaller than 1, or greater than 'last', then those numbers will be silently ignored.

to a central place.

The-Compiler commented:
Quick update: I'm back from my exams and merged the new config - there are at least three PRs related to `:tab-` commands in some way: #2648, #2910 and #2937.

I intend to merge some other PRs today (which don't conflict), and then look at the PR and merge them one-by-one, starting with the oldest one. Please don't rebase on the latest master before that, or you'll just have conflicts again.

Sorry for the pain, but yeah, I knew why I asked people to hold back PRs for a bit :wink:

gilbertw1 commented:
Hah no worries!

I intentionally waited to add support for indexes to some of the other tab commands until the dust has settled. Just ping me here once your you've gotten the other tab related PR's sorted out and I'll do the rebase and add index support to the other commands.

The-Compiler commented:
@Enzime @jgkamat Is any `:tab-*` related stuff still pending? Is anyone of you still working on the `:buffer` → `:tab-focus` merge?

jgkamat commented:
@The-Compiler, I haven't started working working on that at all yet, if @gilbertw1 is ready to work on this now, I'm fine with waiting a bit more, I'm going to be a bit busy in the next 2 weeks, so I might not be able to start on this until after that anyway :P. 

Looking at the diff, I don't think it matters too much, since I don't think this PR touches `:buffer` or `:tab-focus`. I'll make sure to post somewhere when I start working on that so there's no confusion :)

gilbertw1 commented:
Hey guys,

I've started looking back into this and I'm looking for a little advice on how to best integrate with the new tab functions. First of all, adding rich index support to `:tab-clone`, `:tab-close`, `:tab-pin`, and `:tab-only` will be relatively straightforward. However, adding this support to `:tab-give` and `:tab-take` will be a little more nuanced, primarily because we now have to account for windows in addition to tabs.

#### `:tab-take`

`:tab-take` currently takes a single `<window-id>/<index>` combo. eg. `:tab-take 0/2`. We could support this behavior by allowing indexes to be defined using this same notation: `0/1,0/3-0/7,1/2/-1/4` or something similar. This seems reasonable to me and it'd be easy for the user to know which window is which based on the initial tab completion.

#### `:tab-give`

`:tab-give` looks as though it'd be a little more tricky. It currently only takes a window-id and just gives the current tab to that window, eg. `:tab-give 1`. It uses window id completions on the first argument to select the window to give the current tab to.

One possibility would be that we could change `:tab-give` to take two parameters. The first would be an `indexes` parameter that could include 1 or more indexes and it's completions would be all the available tabs, with the second parameter being a `window-id` parameter that would have completions for all the open windows. The drawback here is that the `:tab-give` command would not be as fluid requiring you to select a tab before selecting a window as the second parameter. Whether or not the tabs selected to give should be scoped to the current window or selected from any window (eg. `<window-id>/<tab-index>`) is up for debate. The defaults for both parameters if not provided would be: `indexes -> current-index`, `window-id -> new-window`.

Another possibility that I like a little better (sorry stream of consiousness here), would be to take the above approach, but reverse the order of the parameters, so that the `window-id` would be the first paremeter and the `tab-index(es)` would be the second parameter. This way the command could be used exactly as it is today, without requiring a user to first select an index before selecting a window. The drawback to this approach would be that we'd need to define some placeholder value to enter for a "new" window. Maybe even the string "new" itself would work here.


Sorry for the long post. Do you guys have any thoughts on the above approaches?

The-Compiler commented:
> `:tab-take` currently takes a single `<window-id>/<index>` combo. eg. `:tab-take 0/2`. We could support this behavior by allowing indexes to be defined using this same notation: `0/1,0/3-0/7,1/2/-1/4` or something similar. This seems reasonable to me and it'd be easy for the user to know which window is which based on the initial tab completion.

Seems fine to me - note that you'd have to check that a range is from the same window. Also, I'm assuming the extra `/` in `1/2/-1/4` was an accident?

> Another possibility that I like a little better (sorry stream of consiousness here), would be to take the above approach, but reverse the order of the parameters, so that the `window-id` would be the first paremeter and the `tab-index(es)` would be the second parameter. This way the command could be used exactly as it is today, without requiring a user to first select an index before selecting a window. The drawback to this approach would be that we'd need to define some placeholder value to enter for a "new" window. Maybe even the string "new" itself would work here.

With both parameters optional, that placeholder text would only be needed when actually supplying a range of tabs.

LGTM!

Enzime commented:
For specifying a range for a single window, maybe a format like `<win_id>/<start>-<end>` might look better e.g. `:tab-take 5/1-3,1/1`

gilbertw1 commented:
>Also, I'm assuming the extra / in 1/2/-1/4 was an accident?

Yep, definitely an accident.

> With both parameters optional, that placeholder text would only be needed when actually supplying a range of tabs.

Yea, I was referring to that case, eg. `tab-give new 1,2,3`

> For specifying a range for a single window, maybe a format like <win_id>/<start>-<end> might look better e.g. :tab-take 5/1-3,1/1

I can totally get behind this. That will provide less opportunity for the user to accidentally enter invalid data, specifically the case that @The-Compiler highlighted.

Also, I forgot to explicitly mention it, but I'm planning on supporting tab indexes without window ids specified, in which case it'll just refer to the current window. For example:: `1,2,1/2,1/3` would refer to tabs 1&2 from the current window and tabs 2&3 from window 1. With this being said, I've been thinking that it'd be nice to fully support the `<window-id>/<index>` format for all the `tab-*` commands that I'm adding tab indexes to. They'll still work exactly the same if normal indexes (`l,2,3`) are provided, however will also support specifying tabs from other windows. This would make the commands more flexible and consistent.

The-Compiler commented:
@gilbertw1 Did you have time to look at this already? 😄

gilbertw1 commented:
Yep! I'm a little over half way through the refactor. This past week has been much busier than I anticipated, so I'm going a bit slow :)

I hope to have this ready to review in the next day or two.

The-Compiler commented:
How do things look, @gilbertw1? Do you plan to open a new PR with your changes once they're done? If so, I'd like to close this PR to keep things clean :wink:

gilbertw1 commented:
Sorry, being a really bad contributor right now :(

I'll go ahead and close this PR out and I'll create another one once I'm wrapped up. Much of the code has been changed and the feature scope has been widened so it makes sense.

Thanks!

The-Compiler commented:
No worries, and thanks for the update!
